mantidproject/mantid
# Add container scattering to Discus multiple scattering algorithm

**Description of work.**

Add support for scattering from a container to the Discus multiple scattering algorithm. This means that when generating a result for two scatters it will include extra situations in the simulation such as one scatter happening in the sample container and one in the sample.

The code changes involve converting various bits of processing to cope with a list of S(Q,w) instead of a single S(Q,w) for the sample. The algorithm class now has a new member called `m_SQWSs` that is a boost small_vector of S(Q,w) distributions.

**To test:**

Run the following script and check it works. I suggest you do it on a release build so it runs in a short time. It is a variation on one of the usage examples that is part of the documentation for the algorithm. The variation is that the input workspace has an Aluminium container set up as well as a sample. I've also changed the sample shape to be a cylinder instead of a sphere since it's easier to set up a container for a cylinder:
```
# import mantid algorithms, numpy and matplotlib
from mantid import mtd
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

# S(Q) consisting of single spike at q=1
X=[0.99,1.0,1.01]
Y=[0.,100,0.]
Sofq=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")

two_thetas=[]
for i in range(180):
    two_thetas.append(i)

# workspace with single bin centred at k=1 Angstrom-1
ws = CreateSampleWorkspace(WorkspaceType="Histogram",
                           XUnit="Momentum",
                           Xmin=0.5,
                           Xmax=1.5,
                           BinWidth=1.0,
                           NumBanks=len(two_thetas)//4,
                           BankPixelWidth=2,
                           InstrumentName="testinst")

ids = list(range(1,len(two_thetas)+1))
EditInstrumentGeometry(ws,
    PrimaryFlightPath=14.0,
    SpectrumIDs=ids,
    L2=[2.0] * len(two_thetas),
    Polar=two_thetas,
    Azimuthal=[90.0] * len(two_thetas),
    DetectorIDs=ids,
    InstrumentName="testinst")

SetSample(InputWorkspace=ws,
          Geometry={'Shape': 'Cylinder', 'Radius': 1.0, 'Height': 2.0, 'Centre': [0.,0.,0.], 'Axis': [1.0, 0.0, 0.0]},
          Material={'NumberDensity': 0.02, 'AttenuationXSection': 0.0,
                    'CoherentXSection': 0.0, 'IncoherentXSection': 0.0, 'ScatteringXSection': 80.0})

results_group = DiscusMultipleScatteringCorrection(InputWorkspace=ws, StructureFactorWorkspace=Sofq,
                                                   OutputWorkspace="MuscatResultsNoContainer", NeutronPathsSingle=1000,
                                                   NeutronPathsMultiple=10000, ImportanceSampling=True)

Scatter_1_DeltaFunctionNoContainer = CloneWorkspace('MuscatResultsNoContainer_Scatter_1')
Scatter_2_DeltaFunctionNoContainer = CloneWorkspace('MuscatResultsNoContainer_Scatter_2')
DeleteWorkspace('MuscatResultsNoContainer')

SetSample(InputWorkspace=ws,
          Geometry={'Shape': 'Cylinder', 'Radius': 1.0, 'Height': 2.0, 'Centre': [0.,0.,0.], 'Axis': [1.0, 0.0, 0.0]},
          Material={'NumberDensity': 0.02, 'AttenuationXSection': 0.0,
                    'CoherentXSection': 0.0, 'IncoherentXSection': 0.0, 'ScatteringXSection': 80.0},
          ContainerGeometry={'Shape': 'HollowCylinder', 'Height': 2.0,
                  'InnerRadius': 1.0, 'OuterRadius': 1.2, 'Center': [0.,0.,0.], 'Axis': [1.0, 0.0, 0.0]},
          ContainerMaterial={'ChemicalFormula': 'Al',
                  'NumberDensity': 0.1})

results_group = DiscusMultipleScatteringCorrection(InputWorkspace=ws, StructureFactorWorkspace=Sofq,
                                                   OutputWorkspace="MuscatResultsWithContainer", NeutronPathsSingle=1000,
                                                   NeutronPathsMultiple=10000, ImportanceSampling=True)
# Can't index into workspace group by name (yet) so just get the members from the ADS instead
Scatter_1_DeltaFunctionWithContainer = CloneWorkspace('MuscatResultsWithContainer_Scatter_1')
Scatter_2_DeltaFunctionWithContainer = CloneWorkspace('MuscatResultsWithContainer_Scatter_2')
DeleteWorkspace('MuscatResultsWithContainer')

msplot = plotBin('Scatter_2_DeltaFunctionNoContainer',0)
msplot = plotBin('Scatter_2_DeltaFunctionWithContainer',0, window=msplot)
```
The plot should show two simulated intensities that are broadly similar. The intensity with a container is slightly smaller because there's more material in the beam.

Fixes #33694.





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
